enterprise/server/raft/store/store.go

```diff
@@ -61,13 +64,402 @@ const (
+	if updateType != serf.EventUser {
+		return
+	}
+	userEvent, ok := event.(serf.UserEvent)
+	if !ok {
+		return
+	}
+	if userEvent.Name != constants.NodePartitionUsageEvent {
+		return
+	}
+
+	nu := &rfpb.NodePartitionUsage{}
+	if err := proto.Unmarshal(userEvent.Payload, nu); err != nil {
+		return
+	}
+
+	ut.RemoteUpdate(nu)
+}
+
+// RemoteUpdate processes a usage update broadcast by Raft nodes.
+// Note that this also includes data broadcast by the local node.
+func (ut *usageTracker) RemoteUpdate(usage *rfpb.NodePartitionUsage) {
+	ut.mu.Lock()
+	defer ut.mu.Unlock()
+
+	nhid := usage.GetNode().GetNhid()
+	for _, pu := range usage.GetPartitionUsage() {
+		lpu, ok := ut.byPartition[pu.GetPartitionId()]
+		if !ok {
+			log.Warningf("unknown partition %q", pu.GetPartitionId())
+			continue
+		}
+		lpu.RemoteUpdate(nhid, pu)
+	}
+
+	// Propagate the updated usage to the LRU.
+	for _, u := range ut.byPartition {
+		u.lru.UpdateGlobalSizeBytes(u.GlobalSizeBytes())
+	}
+}
+
+// LocalUpdate processes a usage update from a local replica.
+func (ut *usageTracker) LocalUpdate(rangeID uint64, usage *rfpb.ReplicaUsage, partitionUsages []*rfpb.PartitionMetadata) {
+	ut.mu.Lock()
+	defer ut.mu.Unlock()
+
+	ut.byRange[rangeID] = usage
+
+	// Partition usage is only tracked for leased ranges.
+	if !ut.store.haveLease(rangeID) {
+		ut.removeRangePartitions(rangeID)
+		return
+	}
+
+	for _, u := range partitionUsages {
+		ud, ok := ut.byPartition[u.GetPartitionId()]
+		if !ok {
+			log.Warningf("unknown partition %q", u.GetPartitionId())
+			continue
+		}
+		ud.replicas[rangeID] = u
+	}
+
+	// Propagate the updated usage to the LRU.
+	for _, u := range ut.byPartition {
+		u.lru.UpdateLocalSizeBytes(u.LocalSizeBytes())
+	}
+}
+
+func (ut *usageTracker) removeRangePartitions(rangeID uint64) {
+	for _, u := range ut.byPartition {
+		delete(u.replicas, rangeID)
+	}
+}
+
+func (ut *usageTracker) RemoveRange(rangeID uint64) {
+	ut.mu.Lock()
+	defer ut.mu.Unlock()
+
+	delete(ut.byRange, rangeID)
+
+	ut.removeRangePartitions(rangeID)
+}
+
+func (ut *usageTracker) RangeUsages() []*rfpb.ReplicaUsage {
+	ut.mu.Lock()
+	defer ut.mu.Unlock()
+
+	var us []*rfpb.ReplicaUsage
+	for _, u := range ut.byRange {
+		us = append(us, u)
+	}
+	return us
+}
+
+func (ut *usageTracker) computeUsage() *rfpb.NodePartitionUsage {
+	ut.mu.Lock()
+	defer ut.mu.Unlock()
+	nu := &rfpb.NodePartitionUsage{
+		Node: ut.store.NodeDescriptor(),
+	}
+
+	for _, p := range ut.partitions {
+		up := &rfpb.PartitionMetadata{
+			PartitionId: p.ID,
+		}
+		if u, ok := ut.byPartition[p.ID]; ok {
+			// Sum up total partition usage. Other nodes don't need to know
+			// about individual ranges.
+			for _, ru := range u.replicas {
+				up.SizeBytes += ru.GetSizeBytes()
+				up.TotalCount += ru.GetTotalCount()
+			}
+		}
+		nu.PartitionUsage = append(nu.PartitionUsage, up)
+	}
+	return nu
+}
+
+func (ut *usageTracker) broadcastLoop() {
```

Comment:
this looping and gossiping gets really noisy with large numbers of nodes. Could we drive broadcasts based on a change in usage data greater than a certain size?

Reply:
The looping is cheap, but I think reducing the frequency of gossiping is a fair point. I can follow that up in a separate PR, it would be good to get this change merged.

Reply:
SGTM, but I do think this is relatively important to do soon.

Reply:
It'll be in the next PR.